airqo-platform/AirQo-api
# Encrypt device keys, add input validations and refactor

# Encrypt device keys, add more input validation(sanitisation) and improve code quality for the use case of device creation.

**_WHAT DOES THIS PR DO?_**

- [x] Adds encryption to the device keys - in relation to [issue-433](https://github.com/airqo-platform/AirQo-api/issues/433)
- [x] Also resolves [issue-353](https://github.com/airqo-platform/AirQo-api/issues/353) which is about the device naming convention.
- [x] Resolves[ issue-206](https://github.com/airqo-platform/AirQo-api/issues/206) for trimming our query parameters
- [x] At device creation, input fields no longer require `name` as it is automatically generated in the system. A user just provides `long_name` and other respective mandatory fields. More details are in the PR description.
- [x] One can now use the Mongo ID to perform all CRUD of a device.
- [x] Makes a validation hotfix on device creation. If `powerType` and `mountType` are not provided in the request body, no need to validate them.
- [x] Updates the Update Site endpoint - no need to use separating body parameters for site tags when updating -- you can now just use `site_tags `accordingly. So what you send in the body for site_tags will delete/replace any existing site_tags.
- [x] Adds new flags to the Events collection to distinguish data from primary devices and data for testing.
- [x] Works on the refactoring of `create-device`, `create-site` and `create-event` use cases in preparation for unit testing works.

> - Utilise the mongo ID Object for CRUD operations of a device
> - Incorporate a more powerful logging tool and integrate it with Slack
> - Structure code better in preparation for unit tests
> - Updates the paths of the device creation use case, all endpoints no longer have the `/ts` at the end of their paths.
> - Adds more input validations for the device creation, update, view and decrypt endpoints
> - Input validation for create-site use-case

**_WHAT JIRA STORY/TASK IS RELATED TO THIS PR?_**
N/A

**_WHAT IS THE LINK TO THE PR BRANCH_**
https://github.com/airqo-platform/AirQo-api/tree/issue-device-keys

**_HOW DO I TEST OUT THIS PR?_**
```
cd AirQo-api/src/device-registry
npm install
```
`npm run dev-mac` OR `npm run dev-pc`
Run the create device endpoint and observe the way the keys are stored in the database using the GET command

**_WHICH ENDPOINTS SHOULD BE READY FOR TESTING?:_** 
Since there is a lot of refactoring that has gone on, the following endpoints were affected so they need to be tested...
- [x] Create device
POST http://localhost:3000/api/v1/devices?tenant=airQo
```
{
    "long_name": "AirQo refactoring test",
    "latitude": "1.234",
    "longitude": "0.413",
    "visibility": false,
    "device_manufacturer": "AirQo",
    "product_name": "Gen8",
    "ISP": "Airtel",
    "phoneNumber": "1434235252",
    "description": "an AirQo device",
    "device_number": 12414,
    "generation_version":23,
    "generation_count":12

}
```
- [x] Update device
PUT http://localhost:3000/api/v1/devices?id={OBJECT_ID}&tenant={TENANT}
```
{
    "long_name": "Activelli testing device update"
}
```
- [x] View devices
GET http://localhost:3000/api/v1/devices?tenant=airqo
_Optional query params:_
> id
> name
> device_number
> site_id
> device

- [x] Decrypt read and write Keys
POST: http://localhost:3000/api/v1/devices/decrypt
```
{
    "encrypted_key": "{ENCRYPTION_KEY}"
}
```

The ones below are pending tests, not yet ready......

- [x] Add Events
POST: http://localhost:3000/api/v1/devices/events?tenant={TENANT}
The `device` query param is no longer a requirement so you just go ahead and insert for various devices and sensors at the same time...
```
[
   {
    "device": "momomo",
    "day": "2021-07-05",
    "frequency": "raw",
    "time": "2021-07-05T06:00:00.000Z",
    "device_id": "603d6ce732e033360ae4190b",
    "site_id": "60d2b5107e9238a1a8d38c1a",
     "is_primary": false,
     "is_test_data":false,
    "device_number": null,
    "tenant": "kcca",
....
},
.....
]
```
- [ ] Get Events
GET http://localhost:3000/api/v1/devices/events?tenant={TENANT}
- [ ]  Clear device data
- [x] GET QRCode of device
There is an optional query parameter (`include_site`) which can either be "yes" or "no" accordingly. For "yes", the QR code will contain site details accordingly.
 GET http://localhost:3000/api/v1/devices/qrcode?tenant={TENANT}&device_number={DEVICE_NUMBER}
- [x] Refresh Site details
This just regenerates/refreshes any missing site metadata all in one go, automatically in the background.
PUT http://localhost:3000/api/v1/devices/sites/refresh?tenant=airqo&id={SITE_ID}
- [x] Just generate and return Site Metadata using GPS coordinates
This one just generates site information and returns it without storing it anywhere
POST http://localhost:3000/api/v1/devices/sites/metadata?tenant=airqo
_BODY:_
```
{
    "latitude": -0.321261,
    "longitude": 31.761158
}
```
- [x] Update Site
PUT  http://localhost:3000/api/v1/devices/sites?tenant=airqo&id={SITE_ID}
```
{
    "site_tags": [
        "dusty roads",
        "burning rubbish"
    ]
}
```
- [x] New version (v2) of GET Events
With this version, you can retrieve events for all frequencies on the fly, at the time of request
Defaults to` raw` frequency. Valid frequencies include: `hourly`, `daily` and `raw`
GET http://localhost:3000/api/v2/devices/events?tenant=airqo&frequency=daily

**_IS THERE ANY JENKINS CONSOLE LINK FOR CODE COVERAGE AND BUILD INFO?_**
N/A

**_ARE THERE ANY RELATED PRs?_**
N/A
